### PR TITLE
kvclient/rangefeed: allow starting from a frontier

### PR DESCRIPTION
**Pull request by dt**
This extends the rangefeed client to allow starting a from an existing
frontier, rather than just a list of spans. This is particularly useful
when an initial scan is required and may have previously been partially
completed, as it allows resuming the partial scan from the previous
frontier instead of re-scanning all spans.

Release note: none.
Epic: none.

### CODE REVIEW

File: pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go [hunk 1]
```diff
@@ -341,6 +342,64 @@ func TestRangeFeedMock(t *testing.T) {
 		<-done
 		r.Close()
 	})
+	t.Run("initial scan makes progress", func(t *testing.T) {
```
Comment (stevendanna) [inline pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go:345]: I think this test could use a bit of commentary on what it is doing.

**Review by stevendanna — COMMENTED** [review]
Being able to restart from a frontier is going to be a very nice improvement.

One minor typo in the last commit message:

"allow starting a from" -> "allow starting from"

The unit test I would love here is something where the input are various starting frontiers and the expected output is the set of generated span requests, but I realize that might be easier said than done given the way things are carved up.

File: pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go (hunk 1 above)
Comment (dt) [reply pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go]: Done

**Comment by dt** [comment]
> The unit test I would love here is something where the input are various starting frontiers and the expected output is the set of generated span requests, but I realize that might be easier said than done given the way things are carved up.

The test added here was _sort of_ doing that already, in that requested spans would hang if you over-consumed from it. That said, I've extended it a bit more, to assert that the request is indeed for the expected span -- the one for which we'll mark some subspace starting there as complete before failing this iteration. This means the "results" list now also verifies the request we make, for this one specific set of requests / resumes.

IMO we're testing the important thing here: that we _do eventually_ finish all spans, even when we always fail a rangefeed after completing a strict subset (1 in this case). Verifying what spans we request is nice -- and getting at the same thing -- but the important property is that we can complete n items eventually even if we only ever do k < n before the rangefeed is stopped?

File: pkg/kv/kvclient/rangefeed/rangefeed.go [hunk 2]
```diff
@@ -196,23 +196,27 @@ func (f *RangeFeed) Start(ctx context.Context, spans []roachpb.Span) error {
...
 
-	for _, sp := range spans {
-		if _, err := frontier.Forward(sp, f.initialTimestamp); err != nil {
-			frontier.Release() // release whatever was allocated.
-			return err
-		}
+// StartFromFrontier is like Start but allows passing a frontier to reflect any
+// previous progress. The rangefeed takes ownership of the passed frontier until
+// it is closed; the caller must not interact with it until Close returns. The
+// caller remains responsible for releasing the frontier thereafter however.
```
Comment (stevendanna) [inline pkg/kv/kvclient/rangefeed/rangefeed.go:212]: This is hopefully obvious from the signature, but it is also required that the frontier exactly covers the spans you want to follow with the rangefeed.

**Review by stevendanna — APPROVED** [review]
Thanks for adding the additional commentary on the test. Wholeheartedly agree that your test captures the most useful property. 

I left comments representing some of my internal dialog while re-reading this. But, I don't think any of them represent changes.

File: pkg/kv/kvclient/rangefeed/rangefeed.go (hunk 2 above)
Comment (dt) [reply pkg/kv/kvclient/rangefeed/rangefeed.go]: Updated the comment to say so

File: pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go [hunk 3]
```diff
@@ -341,6 +342,79 @@ func TestRangeFeedMock(t *testing.T) {
...
+
+		var runs int
+		done := false
+		// Run our rangefeed up to 5 times, or until it reports completing its
+		// initial scan, simulating a job performing an initial scan restarting.
+		for ; !done && runs < 5; runs++ {
+			// Sanity check that our frontier isn't empty (i.e. wasn't Released).
+			require.Greater(t, frontier.Len(), 0, frontier.String())
+			// Create a rangefeed that will abort on scan error, to give our test a
+			// change to stop it and create a new one, similar to how a real one might
```
[the remark this answers is not in the document]
Comment (dt) [reply pkg/kv/kvclient/rangefeed/rangefeed_mock_test.go]: done

**Comment by dt** [comment]
hm, need to re-run tests under the race detector/race transport.

**Comment by dt** [comment]
blathers backport 24.1

**Pull request merged**